client/src/client.rs

```diff
@@ -236,12 +236,40 @@ impl<B, E, Block, RA> Client<B, E, Block, RA> where
 		self.backend.state_at(*block)
 	}
 
-	/// Given a `BlockId` and a key prefix, return the matching child storage keys in that block.
+	/// Given a `BlockId` and a key prefix, return the matching storage keys in that block.
 	pub fn storage_keys(&self, id: &BlockId<Block>, key_prefix: &StorageKey) -> sp_blockchain::Result<Vec<StorageKey>> {
 		let keys = self.state_at(id)?.keys(&key_prefix.0).into_iter().map(StorageKey).collect();
 		Ok(keys)
 	}
 
+	/// Paged version of `storage_keys`. Only return up to `count` keys after `start_key` or `prefix`.
+	pub fn storage_keys_paged(
+		&self,
+		id: &BlockId<Block>,
+		prefix: &StorageKey,
+		count: u32,
+		start_key: &Option<StorageKey>
+	) -> sp_blockchain::Result<Vec<StorageKey>> {
+		let state = self.state_at(id)?;
+		let mut result = Vec::<StorageKey>::with_capacity(count as usize);
+		let start_key = start_key.as_ref().unwrap_or(prefix);
+		let prefix = &prefix.0[..];
+
+		let mut current_key = start_key.0.clone();
+		while result.len() < (count as usize) {
+			let maybe_next_key = state.next_storage_key(&current_key[..])
+				.map_err(|e| sp_blockchain::Error::from_state(Box::new(e)))?
+				.filter(|v| v.starts_with(prefix));
```

Comment:
I think this condition should be more strict. `next_storage_key` will return values that are ordered, so when we find first one that doesn't share the prefix we should just quit right away.
Current implementation may traverse entire trie if `count` is greater than than the number of keys that share the same prefix, no?

Also have you considered using `self.state_at(id)?.keys(&key_prefix.0)` and only paging from there? How is that method implemented (I guess it allocates currently), perhaps we could just alter it to do what we want (i.e. return an iterator over prefixed fields instead of allocating everything right away).

Reply:
It does break immediately when finding first result that does not start with prefix. the `filter` call is on `Option` not `Vec`.

But modify `keys` to make it return an iterator sounds a better approach, will do this way.